source/grid/grid_refinement.cc

```diff
@@ -231,10 +231,10 @@ GridRefinement::refine_and_coarsen_fixed_number(
           else
             {
               std::nth_element(tmp.begin(),
-                               tmp.begin() + refine_cells,
+                               tmp.begin() + refine_cells - 1,
                                tmp.end(),
                                std::greater<double>());
-              refine(tria, criteria, *(tmp.begin() + refine_cells));
+              refine(tria, criteria, *(tmp.begin() + refine_cells - 1));
```

Comment:
Certainly looks reasonable to me. I see now how it's difficult to take the average value of two values: We only do the partial sort, so determining the adjacent value would require another partial sort.

Incidentally, the existing code is going to blow up if `refine_cells=n_cells`. So what you have clearly looks better.

Can you update the failing tests?